KK_ButtPlugin/ButtplugController.cs

```diff
@@ -102,21 +135,54 @@ IEnumerator RunVibrate()
         {
             while (!flags.isHSceneEnd)
             {
-                if (flags.nowAnimStateName.Equals("OLoop"))
+                var animator = flags.lstHeroine[0].chaCtrl.animBody;
+
+                float vibrationLevel;
+                if (IsOrgasm)
+                {
+                    vibrationLevel = 1.0f;
+                }
+                else if (IsFemale && hscene.hand.useItems.Any(item => item?.kindTouch == HandCtrl.AibuColliderKind.kokan))
+                {
+                    // add touch vibration from any item between the legs
+                    vibrationLevel = Mathf.InverseLerp(0, flags.speedMaxItem, flags.speedItem) * Mathf.Lerp(0.2f, 1.0f, flags.speedCalc);
+                }
+                else if (IsFemale && flags.mode == HFlag.EMode.masturbation && supportedAnimations.Contains(flags.nowAnimStateName))
                 {
-                    DoVibrate(1.0f);
+                    var info = animator.GetCurrentAnimatorStateInfo(0);
+
+                    // masturbation doesn't have player controlled speed, so use animation speed
+                    // pseudo syncs vibration levels with masturbation animations
+                    var speed = Mathf.Sin(Mathf.Lerp(0, Mathf.PI, info.m_NormalizedTime % 1));
+
+                    vibrationLevel = speed;
                 }
-                else if (!supportedModes.Contains(flags.mode) || !supportedAnimations.Contains(flags.nowAnimStateName))
+                else if (
+                    (IsMale && (!supportedModes.Contains(flags.mode) || !supportedAnimations.Contains(flags.nowAnimStateName))) ||
+                    (IsFemale && (!supportedFemaleModes.Contains(flags.mode) || !supportedAnimations.Contains(flags.nowAnimStateName)))
+                )
                 {
                     // stops vibration when not being lewd
-                    DoVibrate(0.0f);
+                    vibrationLevel = 0.0f;
                 }
                 else
                 {
-                    // vibrate based on the intensity of the player
-                    // minimum vibration above 0 exists so you always feel something along with the animation
-                    DoVibrate(Mathf.Lerp(0.2f, 1.0f, flags.speedCalc));
+                    var speed = Mathf.Lerp(0.2f, 1.0f, flags.speedCalc);
+                    var strength = 1.0f;
+
+                    if (ButtPlugin.SyncVibrationWithAnimation.Value)
+                    {
+                        // Simple sin based intensity amplification based on normalized position in looping animation
+                        // Simulates depth based rhythm without having to know precise animation timings.
+                        // As normalized time is based on animation playback, intensity will also be in sync with speed
+                        var info = animator.GetCurrentAnimatorStateInfo(0);
+                        strength = Mathf.Sin(Mathf.Lerp(0, Mathf.PI, info.m_NormalizedTime % 1));

```

Comment:
Subtracting the animation phase from the normalized time before %1 should make this depth based.

Reply:
trying this out.  The change is subtle if noticeable at all.  This might be because vibrators do not have latency and timing similar to strokers, so it's easier to just keep in sync with a basic curve.